### PR TITLE
Added ninja mingw support and improved CommandGeneratorAction support

### LINKED ISSUE #4103: Added "--experimental=ninja" to the command line, but nothing worked
Added "--experimental=ninja" to the command line, but it has no effect, "build.ninja" is not generated.

scons: scons-local-4.3.0
python: 3.8.5(anaconda)
platform: win11
ninja: 1.10.2
test project: https://github.com/RT-Thread/rt-thread.git    bsp: beaglebone

### PR DESCRIPTION
**Pull request by dmoody256**
Discovered some issues with ninja mingw support from #4103, here is a fix and test.

## Contributor Checklist:

* [x] I have created a new test or updated the unit tests to cover the new/changed functionality.
* [x] I have updated `CHANGES.txt` (and read the `README.rst`)
* [ ] I have updated the appropriate documentation

### CODE REVIEW

File: SCons/Tool/ninja/Methods.py [hunk 1]
```diff
@@ -137,6 +137,9 @@ def get_command(env, node, action):  # pylint: disable=too-many-branches
     comstr = get_comstr(sub_env, action, tlist, slist)
     if not comstr:
         return None
+    
+    if isinstance(comstr, SCons.Util.CLVar):
+        comstr = ' '.join(sub_env.subst(comstr, target=tlist, source=slist))
```
Comment (dmoody256) [inline SCons/Tool/ninja/Methods.py]: isn't there a built in way to make a CLVar into a str?

I tried just str(comstr), but the CLVar contained a File node in the list and it was causing exception.

Comment (mwichmann) [reply SCons/Tool/ninja/Methods.py]: It has a `__str__` method, but perhaps it's too simpleminded for that case?
```python
    def __str__(self):
        return ' '.join(self.data)
```

Comment (dmoody256) [reply SCons/Tool/ninja/Methods.py]: yeah that's where the exception happened. join doesn't know what a `File` object is.

Comment (mwichmann) [reply SCons/Tool/ninja/Methods.py]: Maybe try: 
```python
    return ' '.join((str(d) for d in self.data))
```

File: SCons/Tool/ninja/__init__.py [hunk 2]
```diff
@@ -214,7 +214,7 @@ def ninja_generate_deps(env):
 
     # This adds the required flags such that the generated compile
     # commands will create depfiles as appropriate in the Ninja file.
-    if env["PLATFORM"] == "win32":
+    if env["PLATFORM"] == "win32" and env.get('CC') != 'gcc':
```
Comment (bdbaddog) [inline SCons/Tool/ninja/__init__.py]: could you check:

`if 'mingw' in env['TOOLS']:`

File: SCons/Tool/ninja/__init__.py [hunk 3]
```diff
@@ -267,7 +267,8 @@ def ninja_generate_deps(env):
     def robust_rule_mapping(var, rule, tool):
         provider = gen_get_response_file_command(env, rule, tool)
         env.NinjaRuleMapping("${" + var + "}", provider)
-        env.NinjaRuleMapping(env.get(var, None), provider)
+        if isinstance(env.get(var), str):
```
Comment (bdbaddog) [inline SCons/Tool/ninja/__init__.py:275]: Please add a comment as to why here.

File: test/ninja/mingw_command_generator_action.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+try:
+    import ninja
+except ImportError:
+    test.skip_test("Could not find module in python")
+
+_python_ = TestSCons._python_
+_exe = TestSCons._exe
+
+ninja_bin = os.path.abspath(os.path.join(
```
Comment (bdbaddog) [inline test/ninja/mingw_command_generator_action.py:55]: I think you can use `ninja.BIN_DIR` to find path to ninja exe
Or
`ninja._program('ninja',args)`

Comment (bdbaddog) [reply test/ninja/mingw_command_generator_action.py:55]: https://github.com/scikit-build/ninja-python-distributions/blob/master/src/ninja/__init__.py

File: test/ninja/mingw_command_generator_action.py [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+import sys
+
+import TestSCons
+from TestCmd import IS_WINDOWS
+
+test = TestSCons.TestSCons()
+
+if sys.platform not in ('cygwin', 'win32',):
+    test.skip_test("Skipping mingw test on non-Windows platform %s." % sys.platform)
+
```
Comment (bdbaddog) [inline test/ninja/mingw_command_generator_action.py:38]: you probably also need to skip if mingw is not installed/found on the system.

File: test/ninja/mingw_command_generator_action.py (hunk 4 above)
Comment (bdbaddog) [reply test/ninja/mingw_command_generator_action.py:55]:
```
>>> ninja._program('ninja',['--version'])
1.10.2.git.kitware.jobserver-1
0
```

File: SCons/Tool/ninja/Methods.py (hunk 1 above)
Comment (dmoody256) [reply SCons/Tool/ninja/Methods.py]: done

File: SCons/Tool/ninja/__init__.py (hunk 2 above)
Comment (dmoody256) [reply SCons/Tool/ninja/__init__.py]: done in latest commit

File: SCons/Tool/ninja/__init__.py (hunk 3 above)
Comment (dmoody256) [reply SCons/Tool/ninja/__init__.py:275]: added a comment in latest commit

File: test/ninja/mingw_command_generator_action.py (hunk 4 above)
Comment (dmoody256) [reply test/ninja/mingw_command_generator_action.py:55]: used BIN_DIR in latest commit

File: test/ninja/mingw_command_generator_action.py (hunk 5 above)
Comment (dmoody256) [reply test/ninja/mingw_command_generator_action.py:38]: added in latest commit

File: SCons/Tool/clang.py [hunk 6]
```diff
@@ -81,6 +81,10 @@ def generate(env):
         if match:
             env['CCVERSION'] = match.group(1)
 
+    env['CCDEPFLAGS'] = '.MMD -MF ${TARGET}.d'
```
Comment (dmoody256) [inline SCons/Tool/clang.py]: should be `-MMD`

File: SCons/Tool/gcc.py [hunk 7]
```diff
@@ -57,6 +57,9 @@ def generate(env):
     if version:
         env['CCVERSION'] = version
 
+    env['CCDEPFLAGS'] = '.MMD -MF ${TARGET}.d'
```
Comment (dmoody256) [inline SCons/Tool/gcc.py]: should be `-MMD`

File: SCons/Tool/gxx.py [hunk 8]
```diff
@@ -64,6 +64,9 @@ def generate(env):
     if version:
         env['CXXVERSION'] = version
 
+    env['CCDEPFLAGS'] = '.MMD -MF ${TARGET}.d'
```
Comment (dmoody256) [inline SCons/Tool/gxx.py]: `-MMD`

File: SCons/Tool/clang.py (hunk 6 above)
Comment (bdbaddog) [reply SCons/Tool/clang.py]: Fixed just pushed

File: SCons/Tool/gcc.py (hunk 7 above)
Comment (bdbaddog) [reply SCons/Tool/gcc.py]: fix just pushed

File: SCons/Tool/gxx.py (hunk 8 above)
Comment (bdbaddog) [reply SCons/Tool/gxx.py]: fix just pushed

File: .github/workflows/experimental_tests.yml [hunk 9]
```diff
@@ -30,6 +30,12 @@ jobs:
       # Checks-out your repository under $GITHUB_WORKSPACE, so your job can access it
       - uses: actions/checkout@v2
 
+      - name: Set up MinGW
```
Comment (bdbaddog) [inline .github/workflows/experimental_tests.yml:33]: Was the test failing without mingw installled?

Comment (dmoody256) [reply .github/workflows/experimental_tests.yml:33]: yeah, it found a GCC binary, but that GCC binary was failing, even with just pure scons build. No error messages.

**Comment by dmoody256** [comment]
@bdbaddog this is ready btw

**Pull request merged**